src/sidebar/services/groups.js

```diff
@@ -38,17 +38,67 @@ function groups($rootScope, store, api, isSidebar, localStorage, serviceUrl, ses
     return awaitStateChange(store, mainUri);
   }
 
+  function fetchAnnotation(id) {
+    return api.annotation.get({ id })
+              .then(ann => ann)
+              .catch(() => null);
+  }
+
+  /**
+   * Filter the returned list of groups from the API.
+   *
+   * `filterGroups` performs client-side filtering to hide the "Public" group
+   * for logged-out users under certain conditions.
+   *
+   * @param {Group[]} groups
+   * @param {boolean} isLoggedIn
+   * @param {string|null} directLinkedAnnotation
+   * @return {Promise<Group[]>}
+   */
+  function filterGroups(groups, isLoggedIn, directLinkedAnnotation) {
+    // Logged-in users always see the "Public" group.
+    if (isLoggedIn) {
+      return Promise.resolve(groups);
+    }
+
+    // If the main document URL has no groups associated with it, always show
+    // the "Public" group.
+    var pageHasAssociatedGroups = groups.some(g => g.id !== '__world__');
+    if (!pageHasAssociatedGroups) {
+      return Promise.resolve(groups);
+    }
+
+    // Hide the "Public" group, unless the user specifically visited a direct-
+    // link to an annotation in that group.
+    const removePublicGroup = groups => groups.filter(g => g.id !== '__world__');
```

Comment:
> Curious: why const here but var for pageHasAssociatedGroups? Because it's a function?
 
Oh, just habit. I usually follow the pattern of surrounding code when deciding whether to use `var` (old files) or `const`/`let` (newer files). At some point I think a codebase-wide transformation of `var` => `let`/`const` would make sense to do.

Reply:
Fixed